lib/p2p/Peer.ts

```diff
@@ -265,16 +270,45 @@ class Peer extends EventEmitter {
         resolve();
       };
 
-      const onTimeout = () => {
+      const onError = (err: Error) => {
         cleanup();
-        reject(new Error('Connection timed out.'));
-      };
 
-      this.socket!.once('connect', onConnect);
+        if (!retry) {
+          this.close();
+          reject(err);
+          return;
+        }
+
+        retryDelay = !retryDelay
+          ? Peer.CONNECTION_RETRIES_MIN_DELAY
+          : Math.min(Peer.CONNECTION_RETIES_MAX_DELAY, retryDelay * 2);
+
+        if (Date.now() - startTime + retryDelay > Peer.CONNECTION_RETIES_MAX_PERIOD) {
+          this.close();
```

Comment:
Here too I don't think we need to `close` because we're not connected in the first place.

Reply:
See my comment above.

Before my recent changes, `this.close()` was called even if connection wasn't established yet. I changed that, and that's why i'm calling it here as well.